src/librustc_mir/transform/unreachable_prop.rs

```diff
@@ -0,0 +1,108 @@
+//! A pass that propagates the unreachable terminator of a block to its predecessors
+//! when all of their successors are unreachable. This is achieved through a
+//! post-order traversal of the blocks.
+
+use crate::transform::simplify;
+use crate::transform::{MirPass, MirSource};
+use rustc::mir::*;
+use rustc::ty::TyCtxt;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use std::borrow::Cow;
+
+pub struct UnreachablePropagation;
+
+impl MirPass<'_> for UnreachablePropagation {
+    fn run_pass<'tcx>(
+        &self,
+        _tcx: TyCtxt<'tcx>,
+        _: MirSource<'tcx>,
+        body: &mut BodyAndCache<'tcx>,
+    ) {
```

Comment:
I wonder if we should bail out for optimization level 0. I don't think we should since this optimization is essentially just dead code elimination, which doesn't affect the user's debugging experience.

Other opinions?